File: lib/services/dictionary_service.dart
import 'dart:convert';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:http/http.dart' as http;
import '../data/models/word_definition.dart';

class DictionaryService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  // HÀM LẤY ĐỊNH NGHĨA
  Future<WordDefinition> getDefinition(String text) async {
    final cleanText = text.trim().toLowerCase();

    // 1. Kiểm tra trong Firestore (Cache)
    try {
      final doc =
          await _firestore.collection('dictionary').doc(cleanText).get();
      if (doc.exists) {
        print('✅ Lấy định nghĩa từ Cache (Firestore)');
        return WordDefinition.fromMap(cleanText, doc.data()!);
      }
    } catch (e) {
      print('❌ Lỗi Firestore: $e');
    }

    // 2. Nếu không có, gọi LLM
    print('📡 Đang gọi LLM (Gemini) để lấy định nghĩa...');
    final definition = await _fetchFromLLM(cleanText);

    // 3. Chỉ lưu vào Firestore nếu không phải thông báo lỗi
    final wordDef = WordDefinition.fromLLMResponse(cleanText, definition);

    final isError =
        definition.startsWith("Lỗi") ||
        definition.startsWith("Không thể") ||
        definition.startsWith("AI không tìm thấy");

    if (!isError) {
      try {
        await _firestore
            .collection('dictionary')
            .doc(cleanText)
            .set(wordDef.toMap());
        print('💾 Đã lưu định nghĩa vào Cache');
      } catch (e) {
        print('❌ Không thể lưu vào Firestore: $e');
      }
    } else {
      print('⚠️ Không lưu vào Cache do kết quả bị lỗi.');
    }

    return wordDef;
  }

  // HÀM GỌI API TRÊN VPS
  Future<String> _fetchFromLLM(String text) async {
    try {
      // Gọi API nội bộ của chúng ta thay vì gọi trực tiếp lên Gemini
      final url = Uri.parse(AppStrings.dictionaryEndpoint);
      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({"text": text}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['definition'].toString().trim();
      } else {
        print('⚠️ Lỗi API từ VPS (${response.statusCode}): ${response.body}');
        return "Lỗi: Hệ thống AI đang quá tải, vui lòng thử lại sau.";
      }
    } catch (e) {
      print('❌ Lỗi kết nối mạng: $e');
      return "Lỗi: Không thể kết nối tới máy chủ AI.";
    }
  }
}

File: src/services/dictionaryService.js
import { doc, getDoc, setDoc, serverTimestamp } from 'firebase/firestore';
import { db } from '../firebaseConfig';

const GEMINI_API_KEY = 'AIzaSyAFYx84TkM4LDRNO1EXiGHgylsgLmCWkaI';
const GEMINI_MODELS = ['gemini-1.5-flash', 'gemini-1.5-pro'];

/**
 * Get word definition: cache-first (Firestore) → LLM fallback (Gemini)
 * @param {string} text - Word or phrase to define
 * @returns {Promise<{word: string, definition: string, phonetic: string}>}
 */
export async function getDefinition(text) {
  const cleanText = text.trim().toLowerCase().replace(/[^a-zA-ZÀ-ỹ\s]/g, '');
  if (!cleanText) throw new Error('Empty text');

  // 1. Check Firestore cache
  try {
    const cached = await getDoc(doc(db, 'dictionary', cleanText));
    if (cached.exists()) {
      const data = cached.data();
      return {
        word: data.word || cleanText,
        definition: data.definition || '',
        phonetic: data.phonetic || '',
      };
    }
  } catch (e) {
    console.warn('Cache read failed:', e);
  }

  // 2. Gọi API từ VPS (Giống hệ thống App)
  try {
    const url = 'http://116.118.2.137:8000/api/v1/dictionary';
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text: cleanText }),
    });

    if (!response.ok) throw new Error('Lỗi từ VPS API');

    const json = await response.json();
    if (json.detail) throw new Error(json.detail);
    
    const content = json.definition || '';

    let parsed = {};
    const jsonMatch = content.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      try {
        parsed = JSON.parse(jsonMatch[0]);
      } catch(e) {}
    }

    const definition = parsed.definition || parsed['Nghĩa'] || parsed['Giải thích'] || content.replace(/```json/g, '').replace(/```/g, '').trim();
    const phonetic = parsed.phonetic || parsed['Phiên âm'] || '';

    const result = {
      word: parsed.word || cleanText,
      definition: definition || 'Không tìm thấy định nghĩa',
      phonetic: phonetic,
    };

    // 3. Save to Firestore cache
    try {
      await setDoc(doc(db, 'dictionary', cleanText), {
        ...result,
        updatedAt: serverTimestamp(),
      });
    } catch (e) {}

    return result;
  } catch (vpsError) {
    console.warn(`VPS Dictionary API failed, trying Google Translate fallback:`, vpsError);
    
    // 3. Google Translate Fallback
    try {
      const gtUrl = `https://translate.googleapis.com/translate_a/single?client=gtx&sl=vi&tl=en&dt=t&dt=bd&q=${encodeURIComponent(cleanText)}`;
      const gtRes = await fetch(gtUrl);
      if (!gtRes.ok) throw new Error('Google Translate API failed');
      const gtJson = await gtRes.json();
      
      let translated = '';
      if (gtJson[0] && gtJson[0][0] && gtJson[0][0][0]) {
        translated = gtJson[0][0][0];
      }
      
      let parts = [];
      if (gtJson[1] && gtJson[1].length > 0) {
        gtJson[1].forEach(pos => {
           const type = pos[0] || 'word';
           const terms = (pos[1] || []).join(', ');
           parts.push({ type, terms });
        });
      }
      
      const result = {
        word: cleanText,
        definition: translated, // Primary translation
        phonetic: '',
        parts: parts // Structured parts of speech
      };
      
      try {
        await setDoc(doc(db, 'dictionary', cleanText), {
          ...result,
          updatedAt: serverTimestamp(),
        });
      } catch (e) {}
      
      return result;
    } catch (gtError) {
      console.warn(`Google Translate Fallback failed:`, gtError);
      throw new Error('Không thể tra từ điển lúc này');
    }
  }
}
